src/Lean/Meta/ProofBelow.lean

```diff
@@ -0,0 +1,373 @@
+            let t := 
+              mkAppN vars.indVal[idx] $
+                vars.params ++ vars.motives ++ args[ctx.numParams:] ++ #[hApp]
+            let newDomain ← mkForallFVars xs t
+             
+            withLocalDecl (←copyVarName binder.fvarId!) binder.binderInfo newDomain (k idx)
+          else fail ()
+        else fail ()
+
+  mkMotiveBinder 
+      (vars : Variables) 
+      (indValIdx : Nat) 
+      (binder : Expr) 
+      (domain : Expr) 
+      {α : Type} (k : Expr → MetaM α) : MetaM α := do
+    forallTelescope domain fun xs t => do
+      t.withApp fun f args => do
+        let name := f.constName!
+        let indVal := ctx.typeInfos[indValIdx]
+        let hApp := mkAppN binder xs
+        let t := mkAppN vars.motives[indValIdx] $ args[ctx.numParams:] ++ #[hApp]
+        let newDomain ← mkForallFVars xs t
+         
+        withLocalDecl (←copyVarName binder.fvarId!) binder.binderInfo newDomain k 
+        
+  copyVarName (oldName : FVarId) : MetaM Name := do
+    let binderDecl ← getLocalDecl oldName
+    mkFreshUserName binderDecl.userName
+
+def mkConstructor (ctx : Context) (i : Nat) (ctor : Name) : MetaM Constructor := do
+  let ctorInfo ← getConstInfoCtor ctor
+  let name := ctor.updatePrefix ctx.proofBelowNames[i]
+  let type ← mkCtorType ctx i ctorInfo
+  return { 
+    name := name
+    type := type }
+
+def mkInductiveType 
+  (ctx : Context) 
+  (i : Fin ctx.typeInfos.size) 
+  (indVal : InductiveVal) : MetaM InductiveType := do
+  return {
+    name := ctx.proofBelowNames[i]
+    type := ctx.headers[i]
+    ctors := ←indVal.ctors.mapM (mkConstructor ctx i) }
+
+def mkProofBelowDecl (ctx : Context) : MetaM Declaration := do
+  let lparams := ctx.typeInfos[0].levelParams
+  Declaration.inductDecl 
+    lparams 
+    (ctx.numParams + ctx.motives.size)
+    (←ctx.typeInfos.mapIdxM $ mkInductiveType ctx).toList
+    ctx.typeInfos[0].isUnsafe
+
+partial def proveBrecOn (ctx : Context) (indVal : InductiveVal) (type : Expr) : MetaM Expr := do
+  let main ← mkFreshExprSyntheticOpaqueMVar type
+  let (m, vars) ← intros main.mvarId!
+  let [m] ← applyIH m vars | 
+    throwError "applying the induction hypothesis should only return one goal"
+  let ms ← induction m vars
+  let ms ← applyCtors ms 
+  ms.forM (closeGoal vars)
+  instantiateMVars main
+where
+  intros (m : MVarId) : MetaM (MVarId × BrecOnVariables) := do
+    let (params, m) ← introNP m indVal.numParams
+    let (motives, m) ← introNP m ctx.motives.size
+    let (indices, m) ← introNP m indVal.numIndices
+    let (witness, m) ← intro1P m
+    let (indHyps, m) ← introNP m ctx.motives.size
+    return (m, ⟨params, motives, indices, witness, indHyps⟩)
+
+  applyIH (m : MVarId) (vars : BrecOnVariables) : MetaM (List MVarId) := do
+    match ←vars.indHyps.findSomeM? 
+      (fun ih => do try some $ (←apply m $ mkFVar ih) catch ex => none) with
+    | some goals => goals
+    | none => throwError "cannot apply induction hypothesis: {←ppGoal m}"
```

Comment:
```
throwError "cannot apply induction hypothesis:\n{MessageData.ofGoal m}"
```
We don't eagerly pretty print goals. See commend about `ppExpr`.

Reply:
done